File: modules/wem-api/src/main/java/com/enonic/wem/api/schema/content/form/inputtype/InputTypeName.java
package com.enonic.wem.api.schema.content.form.inputtype;


public class InputTypeName
{
    private final String ref;

    private final String name;

    private final boolean custom;

    public static InputTypeName from( final String s )
    {
        if ( s.startsWith( "custom:" ) )
        {
            return new InputTypeName( s.substring( "custom:".length(), s.length() ), true );
        }
        else
        {
            return new InputTypeName( s, false );
        }
    }

    public static InputTypeName from( final InputType inputType )
    {
        return InputTypeName.from( inputType.getName() );
    }

    public InputTypeName( final String name, final boolean custom )
    {
        this.name = name;
        this.custom = custom;
        this.ref = custom ? "custom:" + name : "" + name;
    }

    public String getName()
    {
        return this.name;
    }

    public boolean isCustom()
    {
        return this.custom;
    }

    public String toString()
    {
        return ref;
    }
}

File: modules/admin-ui/src/main/resources/web/admin/common/js/form/InputTypeName.ts
module api.form {

    export class InputTypeName implements api.Equitable {

        private static CUSTOM_PREFIX: string = "custom:";

        private custom: boolean;

        private name: string;

        private refString: string;

        static parseInputTypeName(str: string) {
            if (str.substr(0, InputTypeName.CUSTOM_PREFIX.length) == InputTypeName.CUSTOM_PREFIX) {
                return new InputTypeName(str.substr(InputTypeName.CUSTOM_PREFIX.length, str.length), true);
            }
            else {
                return new InputTypeName(str, false);
            }
        }

        constructor(name: string, custom: boolean) {
            this.name = name;
            this.custom = custom;

            if (this.custom) {
                this.refString = InputTypeName.CUSTOM_PREFIX + name;
            }
            else {
                this.refString = name;
            }
        }

        getName(): string {
            return this.name;
        }

        isBuiltIn(): boolean {
            return !this.custom;
        }

        toString(): string {
            return this.refString;
        }

        public toJson(): string {

            return this.toString()
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, InputTypeName)) {
                return false;
            }

            var other = <InputTypeName>o;

            if (!api.ObjectHelper.booleanEquals(this.custom, other.custom)) {
                return false;
            }

            if (!api.ObjectHelper.stringEquals(this.name, other.name)) {
                return false;
            }

            return true;
        }
    }
}
